kong/clustering/services/sync/rpc.lua

```diff
@@ -0,0 +1,324 @@
+    }, { ttl = purge_delay })
+    if not ok then
+      ngx_log(ngx_ERR, "unable to update clustering data plane status: ", err)
+    end
+
+    -- is the node empty? If so, just do a full sync to bring it up to date faster
+    if default_namespace_version == 0 or
+       self.strategy:get_latest_version() - default_namespace_version > FULL_SYNC_THRESHOLD
+    then
+      -- we need to full sync because holes are found
+
+      ngx_log(ngx_INFO,
+              "[kong.sync.v2] database is empty or too far behind for node_id: ", node_id,
+              ", current_version: ", default_namespace_version,
+              ", forcing a full sync")
+
+
+      local deltas, err = declarative.export_config_sync()
+      if not deltas then
+        return nil, err
+      end
+
+      -- wipe dp lmdb, full sync
+      return gen_delta_result(deltas, true)
+    end
+
+    local res, err = self.strategy:get_delta(default_namespace_version)
+    if not res then
+      return nil, err
+    end
+
+    if #res == 0 then
+      ngx_log(ngx_DEBUG,
+              "[kong.sync.v2] no delta for node_id: ", node_id,
+              ", current_version: ", default_namespace_version,
+              ", node is already up to date" )
+      return gen_delta_result(res, false)
+    end
+
+    -- some deltas are returned, are they contiguous?
+    if res[1].version == default_namespace.version + 1 then
+      -- doesn't wipe dp lmdb, incremental sync
+      return gen_delta_result(res, false)
+    end
+
+    -- we need to full sync because holes are found
+    -- in the delta, meaning the oldest version is no longer
+    -- available
+
+    ngx_log(ngx_INFO,
+            "[kong.sync.v2] delta for node_id no longer available: ", node_id,
+            ", current_version: ", default_namespace_version,
+            ", forcing a full sync")
+
+    local deltas, err = declarative.export_config_sync()
+    if not deltas then
+      return nil, err
+    end
+
+    -- wipe dp lmdb, full sync
+    return gen_delta_result(deltas, true)
+  end)
+end
+
+
+function _M:init_dp(manager)
+  -- DP
+  -- Method: kong.sync.v2.notify_new_version
+  -- Params: new_versions: list of namespaces and their new versions, like:
+  -- { { namespace = "default", new_version = 1000, }, }
+  manager.callbacks:register("kong.sync.v2.notify_new_version", function(node_id, new_versions)
+    -- currently only default is supported, and anything else is ignored
+    for namespace, new_version in pairs(new_versions) do
+      if namespace == "default" then
+        local version = new_version.new_version
+        if not version then
+          return nil, "'new_version' key does not exist"
+        end
+
+        local lmdb_ver = tonumber(declarative.get_current_hash()) or 0
+        if lmdb_ver < version then
+          return self:sync_once()
+        end
+
+        return true
+      end
+    end
+
+    return nil, "default namespace does not exist inside params"
+  end)
+end
```

Comment:
Is it possible for the value of `new_versions` to contain multiple entries for the `default` namespace, like this?

```lua
{
  { namespace = "default", new_version = 1 },
  { namespace = "default", new_version = 3 },
}
```

If so, then this logic might be faulty. If `lmdb_ver = 2`, then we will exit the loop after one iteration (because `lmdb_ver (2) < version (1)`) instead of evaluating the next entry which has `new_version = 3` and would trigger a sync operation.

Reply:
@flrgh This kind of scenarios will not be expected, I would consider making it:

`{ default = 3 }` instead to avoid this kind of errors. WDTY?

Reply:
My mistake, I wasn't picturing the shape of the data correctly. According to the code, here's the expected shape of `new_versions`:

```
{
  default = { new_version = 1 },
}
```

Because the namespace is also the table field/key, it shouldn't be possible to end up in this code with more than one `default` namespace.